### PR TITLE
Add linsolve choice.

### PR DESCRIPTION
**Pull request by pkofod**

### CODE REVIEW

**Comment by pkofod** [comment]
Seems to work fine with `gmres!` and `idrs!`. Other examples from other (?) packages are welcome as well.

**Comment by ChrisRackauckas** [comment]
Is it possible to swap out the Jacobian type? I would like to see a test that does gmres with a matrix-free representation of the Jacobian given by a LinearMaps.jl FunctionMap.

File: test/linsolve.jl [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+    res_default_solve = nlsolve(df, prob[2]; method=:newton)
+    prob = rosenbrock()
+    df = prob[1]
+    res_gmres_solve = nlsolve(df, prob[2]; method=:newton, linsolve = gmres!)
+    @test res_gmres_solve.zero ≈ res_default_solve.zero
+    res_idrs_solve = nlsolve(df, prob[2]; method=:newton, linsolve = idrs!)
+    @test res_idrs_solve.zero ≈ res_default_solve.zero
+
+    function loud_solve(x, A, b)
+        println("linear solver is passed trough")
```
[the remark this answers is not in the document]
Comment (pkofod) [reply test/linsolve.jl]: I should.

**Comment by pkofod** [comment]
> Is it possible to swap out the Jacobian type? I would like to see a test that does gmres with a matrix-free representation of the Jacobian given by a LinearMaps.jl FunctionMap.

That should be possible, yes. Let me see if I can find the appropriate types you're talking about and come up with an example (if you have a link handy that shows how it works I'd be happy to see it)

**Comment by pkofod** [comment]
> That should be possible, yes. Let me see if I can find the appropriate types you're talking about and come up with an example (if you have a link handy that shows how it works I'd be happy to see it)

Yeah it probably isn't ready right now. The problem is that some internals assume `.=` will work, and I suspect you're asking for some sort of immutable storage type, righT?

**Comment by ChrisRackauckas** [comment]
oh okay. But things like sparse work?

**Comment by pkofod** [comment]
If you’re willing to reuse the object, sure. What sparse problem could benefit from this? Are you thinking calling some nonjulia library?

**Comment by pkofod** [comment]
This works so I'm putting it in. I'm going to collect some examples and maybe make some corrections to better handle immutable jacobians before tagging a version.

**Pull request merged**